ocrd/ocrd/resource_manager.py

```diff
@@ -0,0 +1,142 @@
+from pathlib import Path
+import re
+from shutil import copyfileobj, copytree
+from tempfile import TemporaryFile
+from tarfile import open as open_tarfile
+
+import requests
+from yaml import safe_load
+
+from .constants import RESOURCE_LIST_FILENAME
+
+from ocrd_validators import OcrdResourceListValidator
+from ocrd_utils import getLogger
+from ocrd_utils.constants import HOME, XDG_CACHE_HOME
+from ocrd_utils.os import list_resource_candidates, list_all_resources, pushd_popd
+
+builtin_list_filename = Path(RESOURCE_LIST_FILENAME)
+user_list_filename = Path(HOME, 'ocrd', 'resources.yml')
+
+class OcrdResourceManager():
+
+    """
+    Managing processor resources
+    """
+    def __init__(self):
+        self.log = getLogger('ocrd.resource_manager')
+        self.database = {}
+        self.load_resource_list(builtin_list_filename)
+        self.load_resource_list(user_list_filename)
+
+    def load_resource_list(self, list_filename):
+        if list_filename.is_file():
+            with open(list_filename, 'r', encoding='utf-8') as f:
+                list_loaded = safe_load(f)
+            report = OcrdResourceListValidator.validate(list_loaded)
+            if not report.is_valid:
+                self.log.error('\n'.join(report.errors))
+                raise ValueError("Resource list %s is invalid!" % (list_filename))
+            for executable, resource_list in list_loaded.items():
+                if executable not in self.database:
+                    self.database[executable] = []
+                # Prepend, so user provided is sorted before builtin
+                self.database[executable] = list_loaded[executable] + self.database[executable]
+
+    def list_available(self, executable=None):
+        """
+        List models available for download by processor
+        """
+        if executable:
+            return [(executable, self.database[executable])]
+        return [(x, y) for x, y in self.database.items()]
+
+    def list_installed(self, executable=None):
+        """
+        List installed resources, matching with registry by ``name``
+        """
+        ret = []
+        for executable in [executable] if executable else self.database.keys():
+            reslist = []
+            for res_filename in list_all_resources(executable):
+                res_name = Path(res_filename).name
+                resdict = [x for x in self.database[executable] if x['name'] == res_name]
+                if not resdict:
+                    # TODO handle gracefully
```

Comment:
I'm confused: What exact circumstances would lead us into this branch? 

My test already fails earlier: After…

    ocrd resmgr download -n ocrd-cis-ocropy-recognize https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz

…I still get _no_ results for…

    ocrd resmgr list-installed

…(because the unregistered download does not touch the user resources – see other comment – and the outer loop only looks for registered ones), but…

    ocrd resmgr list-installed -e ocrd-cis-ocropy-recognize

…already fails with…

```
  File "venv/lib/python3.6/site-packages/ocrd/cli/resmgr.py", line 50, in list_installed
    for executable, reslist in resmgr.list_installed(executable):
  File "venv/lib/python3.6/site-packages/ocrd/resource_manager.py", line 62, in list_installed
    resdict = [x for x in self.database[executable] if x['name'] == res_name]
KeyError: 'ocrd-cis-ocropy-recognize'
```

So it would seem that you have to enter either via a loaded and validated resource database (in which case IIUC there will always be an entry for any of its resources), or just by name (in which case IIUC no database lookup is warranted anyway)...

Reply:
> …I still get no results for…

The problem is: what do we iterate? Do we iterate over all the items in all the XDG_FOLDERS that follow our convention (`ocrd-.*`) or over the keys in the `resource_list.yml`? I chose the latter for simplicity's sake but I understand how that can be confusing.

The branch is triggered if there is a resource for one of the processors for which we know there are resources (has a key in `resource_list.yml`) but not the resource with this particular name.

I'll change the outer loop to actually inspect the file system.

Reply:
79116038